Changes to accept every Dataview metadata format

theofbonin opened the pull request:
Changes to also parse for "(field:: value)" and "[field:: value]",
as opposed to only "field:: value".

chhoumann commented on src/parser.ts:
```diff
@@ -42,20 +42,18 @@ export default class MetaEditParser {
-
-            return obj;
-        },  []);
+        const regex = /([^\n\r\(\[]*)::\s?([^\)\]\n\r]*)/g;
```
```suggestion
        const regex = /[\[\(]([^\n\r\(\[]*)::\s?([^\)\]\n\r]*)[\]\)]/g;
```

chhoumann reviewed (changes requested):
Hey @theofbonin!

Thanks for the PR!

I've been testing for a bit using the Dataview docs material on inline fields:
```markdown
Basic Field:: Some random Value
**Bold Field**:: Nice!
I would rate this a [rating:: 9]! It was [mood:: acceptable].
- [ ] Send an mail to David about the deadline [due:: 2022-04-05].
This will not show the (longKeyIDontNeedWhenReading:: key).
```

I get the following results:
![image](https://user-images.githubusercontent.com/29108628/222356160-f97695fd-80d4-4dce-b72a-8641d8e327d5.png)

And zooming in on `rating` gives
![image](https://user-images.githubusercontent.com/29108628/222356231-871b470a-1bc9-4d91-9c13-8703dc6e9da8.png)

Which does seem to provide unexpected results. I'd expect it to prompt me for just `rating` with a pre-filled value of `9`, rather than splitting the line.

Regarding your regex in `parseInlineFields`: I'd use a `*` after the `\s`  rather than `?` so the second match group doesn't get spaces, in case there are more than one. E.g. `[mood:     acceptable]` would ignore the many spaces.
![image](https://user-images.githubusercontent.com/29108628/222357142-3a369af3-97b7-4dc3-8e5b-6cbf9d411ea9.png)

It also seems to be the cause of the unexpected results I mentioned above. Using the following leads to normal results:
Using `const regex = /[\[\(]([^\n\r\(\[]*)::\s?([^\)\]\n\r]*)[\]\)]/g;`
The only thing I added was `[\[\(]` before the first group (key) and `[\]\)]` after the second group (value).
Result:
![image](https://user-images.githubusercontent.com/29108628/222358983-3367fd1e-9ba9-4519-80ea-9d68285d1359.png)

With that done, it seems to work from my tests! :)

theofbonin commented:
Hey @chhoumann!

I made some changes and now it seems to be working for every case. I fixed some bugs that I found and made it work with some other cases that I had thought about.

<img width="596" alt="Screenshot 2023-03-02 at 09 53 04" src="https://user-images.githubusercontent.com/69220259/222449316-07ae5ace-fc6f-4385-a4e2-c5bc74a7796b.png">
<img width="680" alt="Screenshot 2023-03-02 at 09 53 50" src="https://user-images.githubusercontent.com/69220259/222449327-0210b258-fd7f-42e9-b722-89fdd4175eb2.png">
<img width="908" alt="Screenshot 2023-03-02 at 10 00 55" src="https://user-images.githubusercontent.com/69220259/222449332-8f47c4ac-d315-4b58-8ecb-2588a0640658.png">

I changed the regex in "parser.ts" to a new one that conforms to the review:
`/[\[\(]?([^\n\r\(\[]*)::[ ]*([^\)\]\n\r]*)[\]\)]?/g`
Instead of using `\s*`, I used `[ ]*`, because using `\s*` would cause problems with properties that are declared without a value and with a newline right in front of the `::` (for example, field::\n).

I had to create the "escapeSpecialCharacters" method because properties with special regex characters were causing problems when editing the value. For instance, **Bold Field** was causing:

`Uncaught (in promise) SyntaxError: Invalid regular expression: /**Bold Field**:{1,2}/`

I changed the regexes with `${property.key}` to `${this.escapeSpecialCharacters(property.key)}`.

I also changed the regexes in "metaController.ts" to conform to the review.

Known issues:
When there is more than one of the same field on the same file, it appears many times on the Metaedit screen (you can see this on the second screenshot with the field `mood`). This shouldn't be an issue because the "right" way to declare fields on dataview is using lists or arrays: 
```
field1:: [value1, value2, value3]
field2::
    - value1
    - value2
    - ...
```
Even though it does accept many of the same field, like:
```
field:: value1
field:: value2
field:: value3
```
This will be considered by dataview as `field:: [value1, value2, value3]`. Another problem related to this same thing is that when you modify the value of a field, it changes all occurrences of it in the file.

This issue was already happening even without accepting `[field:: value]` and `(field:: value)`, so it is not really related. However, I could try to figure out a way to work with many of the same field in the same file if you think that would be necessary.

chhoumann commented:
This is fantastic, @theofbonin! Everything looks great. Awesome job. You even fixed #54! Thank you for the PR!
Tagging here to close: 
fix #54 

I'd love to see issues where you document your findings on the known issues you mention. I think those should be fixed, but I agree that this PR isn't the place to do it. It's good to avoid bloating change requests with too many things at once.
